File: create/qa/audit_bundle_completeness.go
package qa

import (
	"fmt"
	"path/filepath"
	"regexp"
	"strings"
	"sync"

	"github.com/nyamiiko/script_hub/go_scripts/hub"
	"github.com/nyamiiko/script_hub/go_scripts/pipeline"
)

var scriptLabelRe = regexp.MustCompile(`(?i)^(.+?)\s*=\s*type=`)
var mitmBadTokens = map[string]bool{"%INSERT%": true, "%APPEND%": true}

type bundleSpec struct {
	label                        string
	output                       string
	sources                      []hub.SourceSpec
	requiredScriptLabels         []string
	requiredMitmHosts            []string
	requiredScriptPathSubstrings []string
}

func getScriptLabels(sections []hub.ModuleSection) map[string]bool {
	labels := make(map[string]bool)
	for _, section := range sections {
		name := section.Name
		if name != "Script" {
			continue
		}
		lines := section.Lines
		for _, line := range lines {
			m := scriptLabelRe.FindStringSubmatch(strings.TrimSpace(line))
			if len(m) > 1 {
				labels[strings.TrimSpace(m[1])] = true
			}
		}
	}
	return labels
}

func getMitmHosts(sections []hub.ModuleSection) map[string]bool {
	hosts := make(map[string]bool)
	for _, section := range sections {
		name := section.Name
		if name != "MITM" {
			continue
		}
		lines := section.Lines
		for _, line := range lines {
			line = strings.TrimSpace(line)
			if !strings.HasPrefix(strings.ToLower(line), "hostname") {
				continue
			}
			part := regexp.MustCompile(`(?i)^hostname\s*=\s*`).ReplaceAllString(line, "")
			part = regexp.MustCompile(`(?i)^(%APPEND%|%INSERT%)\s*`).ReplaceAllString(part, "")
			for _, token := range strings.Split(part, ",") {
				token = strings.TrimSpace(token)
				if token != "" && !mitmBadTokens[token] {
					hosts[token] = true
				}
			}
		}
	}
	return hosts
}

func auditBundle(spec bundleSpec) []string {
	var issues []string
	if !hub.ValidateFileExists(spec.output, "") {
		return []string{fmt.Sprintf("%s: bundle missing at %s", spec.label, spec.output)}
	}

	bundleText := hub.ReadFileString(spec.output)
	_, bundleSections := hub.ParseModule(bundleText)
	bundleScripts := getScriptLabels(bundleSections)
	bundleMitm := getMitmHosts(bundleSections)

	if strings.Contains(bundleText, "%INSERT%") {
		issues = append(issues, fmt.Sprintf("%s: contains invalid MITM token %%INSERT%%", spec.label))
	}

	for _, needle := range spec.requiredScriptPathSubstrings {
		if !strings.Contains(bundleText, needle) {
			issues = append(issues, fmt.Sprintf("%s: vendored script-path missing: %s", spec.label, needle))
		}
	}

	upstreamScripts := make(map[string]bool)
	upstreamMitm := make(map[string]bool)

	for _, src := range spec.sources {
		text := hub.SafeDownload(src.URL, 2, 60)
		if text == "" {
			issues = append(issues, fmt.Sprintf("%s: failed to fetch upstream [%s]", spec.label, src.Label))
			continue
		}
		_, sections := hub.ParseModule(text)
		for k := range getScriptLabels(sections) {
			upstreamScripts[k] = true
		}
		for k := range getMitmHosts(sections) {
			upstreamMitm[k] = true
		}
	}

	var missingScripts []string
	for k := range upstreamScripts {
		if !bundleScripts[k] {
			missingScripts = append(missingScripts, k)
		}
	}
	if len(missingScripts) > 0 {
		preview := strings.Join(missingScripts, ", ")
		if len(missingScripts) > 12 {
			preview = strings.Join(missingScripts[:12], ", ") + "…"
		}
		issues = append(issues, fmt.Sprintf("%s: missing script label(s): %s", spec.label, preview))
	}

	for _, req := range spec.requiredScriptLabels {
		if !bundleScripts[req] {
			issues = append(issues, fmt.Sprintf("%s: required script missing: %s", spec.label, req))
		}
	}

	missingMitmPositive := make(map[string]bool)
	for k := range upstreamMitm {
		if !bundleMitm[k] && !strings.HasPrefix(k, "-") {
			missingMitmPositive[k] = true
		}
	}
	for _, req := range spec.requiredMitmHosts {
		if !bundleMitm[req] {
			missingMitmPositive[req] = true
		}
	}

	if len(missingMitmPositive) > 0 {
		var missingArr []string
		for k := range missingMitmPositive {
			missingArr = append(missingArr, k)
		}
		issues = append(issues, fmt.Sprintf("%s: MITM missing host(s): %s", spec.label, strings.Join(missingArr, ", ")))
	}

	return issues
}

func getSpecs() []bundleSpec {
	return []bundleSpec{
		{
			label:             "BiliBili",
			output:            filepath.Join(hub.MODULES_DIR, "surge/amplify_nexus/📺 BiliBili增强合集.sgmodule"),
			sources:           pipeline.BilibiliSources,
			requiredMitmHosts: []string{"app.bilibili.com"},
		},
		{
			label:   "Apple",
			output:  filepath.Join(hub.MODULES_DIR, "surge/amplify_nexus/🍎 Apple服务增强合集.sgmodule"),
			sources: pipeline.AppleSources,
		},
		{
			label:             "Panel utilities",
			output:            filepath.Join(hub.MODULES_DIR, "surge/amplify_nexus/📊 面板工具合集.sgmodule"),
			sources:           pipeline.UtilitiesSources,
			requiredMitmHosts: []string{"net-lsp-x.com"},
		},
		{
			label:   "Script Hub devtools",
			output:  filepath.Join(hub.MODULES_DIR, "surge/amplify_nexus/🧰 Script Hub 配套工具合集.sgmodule"),
			sources: pipeline.DevtoolsSources,
			requiredScriptLabels: []string{
				"Sub-Store Core",
				"Sub-Store Simple",
				"{{{sync}}}",
				"{{{produce}}}",
				"Rewrite: BoxJs",
				"Script Hub: 前端",
			},
			requiredMitmHosts: []string{"sub.store", "script.hub", "boxjs.com"},
			requiredScriptPathSubstrings: []string{
				"modules/source/scripts/github_com_sub-store-org_sub-store-1.min.js",
				"modules/source/scripts/raw_githubusercontent_com_f59ef7_script-hub.js",
			},
		},
		{
			label:   "Weibo (PROMAX source)",
			output:  filepath.Join(hub.ROOT, "rulesets/Sources/LocalModules/Weibo.ADBlock.sgmodule"),
			sources: pipeline.WeiboSources,
		},
	}
}

func auditPromaxModules() []string {
	var issues []string
	rels := []string{
		"modules/surge/head_expanse/🚫 Universal Ad-Blocking Rules Dependency Component PROMAX (Kali-style).sgmodule",
		"modules/surge/head_expanse/📱 Universal Ad-Blocking Rules (PROMAX Lite).sgmodule",
	}
	for _, rel := range rels {
		path := filepath.Join(hub.ROOT, rel)
		if !hub.ValidateFileExists(path, "") {
			issues = append(issues, fmt.Sprintf("PROMAX: missing %s", rel))
			continue
		}
		text := hub.ReadFileString(path)
		if strings.Contains(text, "%INSERT%") {
			issues = append(issues, fmt.Sprintf("PROMAX: %%INSERT%% found in %s", filepath.Base(path)))
		}
	}
	return issues
}

func RunAuditBundleCompleteness() int {
	specs := getSpecs()
	hub.Section("Bundle completeness audit")
	
	var wg sync.WaitGroup
	var mu sync.Mutex
	var allIssues []string

	for _, spec := range specs {
		wg.Add(1)
		go func(s bundleSpec) {
			defer wg.Done()
			hub.Info(fmt.Sprintf("Checking %s…", s.label))
			issues := auditBundle(s)
			if len(issues) > 0 {
				mu.Lock()
				allIssues = append(allIssues, issues...)
				mu.Unlock()
			}
		}(spec)
	}

	wg.Add(1)
	go func() {
		defer wg.Done()
		issues := auditPromaxModules()
		if len(issues) > 0 {
			mu.Lock()
			allIssues = append(allIssues, issues...)
			mu.Unlock()
		}
	}()

	wg.Wait()

	if len(allIssues) > 0 {
		hub.Error(fmt.Sprintf("Found %d issue(s):", len(allIssues)))
		for _, item := range allIssues {
			fmt.Printf("  • %s\n", item)
		}
		return 1
	}

	hub.Success(fmt.Sprintf("All %d bundle(s) + PROMAX passed completeness audit.", len(specs)))
	return 0
}

File: create/scripts/qa/audit_bundle_completeness.py
#!/usr/bin/env python3
"""Audit merged Surge bundle modules for merge regressions (MITM, script dedupe)."""

from __future__ import annotations

import re
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Sequence

SCRIPTS_DIR = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(SCRIPTS_DIR))

from hub.common import Logger, get_project_root
from hub.merge_upstream import _fetch_module
from hub.module_sanitizer import parse_module

SCRIPT_LABEL_RE = re.compile(r"^(.+?)\s*=\s*type=", re.I)
MITM_BAD_TOKENS = frozenset({"%INSERT%", "%APPEND%"})


@dataclass
class BundleSpec:
    label: str
    output: Path
    sources: Sequence[tuple[str, str]]
    required_script_labels: Sequence[str] = ()
    required_mitm_hosts: Sequence[str] = ()
    required_script_path_substrings: Sequence[str] = ()


def _script_labels(sections: list[tuple[str, list[str]]]) -> set[str]:
    labels: set[str] = set()
    for name, lines in sections:
        if name != "Script":
            continue
        for line in lines:
            m = SCRIPT_LABEL_RE.match(line.strip())
            if m:
                labels.add(m.group(1).strip())
    return labels


def _mitm_hosts(sections: list[tuple[str, list[str]]]) -> set[str]:
    hosts: set[str] = set()
    for name, lines in sections:
        if name != "MITM":
            continue
        for line in lines:
            if not line.strip().lower().startswith("hostname"):
                continue
            part = re.sub(r"^hostname\s*=\s*", "", line.strip(), flags=re.I)
            part = re.sub(r"^(%APPEND%|%INSERT%)\s*", "", part, flags=re.I)
            for token in part.split(","):
                token = token.strip()
                if token and token not in MITM_BAD_TOKENS:
                    hosts.add(token)
    return hosts


def audit_bundle(spec: BundleSpec) -> list[str]:
    issues: list[str] = []
    if not spec.output.is_file():
        return [f"{spec.label}: bundle missing at {spec.output}"]

    bundle_text = spec.output.read_text(encoding="utf-8")
    _, bundle_sections = parse_module(bundle_text)
    bundle_scripts = _script_labels(bundle_sections)
    bundle_mitm = _mitm_hosts(bundle_sections)

    if "%INSERT%" in bundle_text:
        issues.append(f"{spec.label}: contains invalid MITM token %INSERT%")

    for needle in spec.required_script_path_substrings:
        if needle not in bundle_text:
            issues.append(f"{spec.label}: vendored script-path missing: {needle}")

    upstream_scripts: set[str] = set()
    upstream_mitm: set[str] = set()

    for src_label, url in spec.sources:
        try:
            text = _fetch_module(url, src_label)
        except RuntimeError as exc:
            issues.append(f"{spec.label}: failed to fetch upstream [{src_label}]: {exc}")
            continue
        _, sections = parse_module(text)
        upstream_scripts |= _script_labels(sections)
        upstream_mitm |= _mitm_hosts(sections)

    missing_scripts = upstream_scripts - bundle_scripts
    if missing_scripts:
        issues.append(
            f"{spec.label}: missing script label(s): {', '.join(sorted(missing_scripts)[:12])}"
            + ("…" if len(missing_scripts) > 12 else "")
        )

    for req in spec.required_script_labels:
        if req not in bundle_scripts:
            issues.append(f"{spec.label}: required script missing: {req}")

    missing_mitm_positive = {
        h for h in (upstream_mitm - bundle_mitm) if not h.startswith("-")
    }
    for req in spec.required_mitm_hosts:
        if req not in bundle_mitm:
            missing_mitm_positive.add(req)
    if missing_mitm_positive:
        issues.append(
            f"{spec.label}: MITM missing host(s): {', '.join(sorted(missing_mitm_positive))}"
        )

    return issues


def _specs(root: Path) -> list[BundleSpec]:
    from pipeline import merge_bundles as mb

    return [
        BundleSpec(
            "BiliBili",
            Path(mb.BILIBILI_OUTPUT),
            mb.BILIBILI_SOURCES,
            required_mitm_hosts=("app.bilibili.com",),
        ),
        BundleSpec(
            "Apple",
            Path(mb.APPLE_OUTPUT),
            mb.APPLE_SOURCES,
        ),
        BundleSpec(
            "Panel utilities",
            Path(mb.UTILITIES_OUTPUT),
            mb.UTILITIES_SOURCES,
            required_mitm_hosts=("net-lsp-x.com",),
        ),
        BundleSpec(
            "Script Hub devtools",
            Path(mb.DEVTOOLS_OUTPUT),
            mb.DEVTOOLS_SOURCES,
            required_script_labels=(
                "Sub-Store Core",
                "Sub-Store Simple",
                "{{{sync}}}",
                "{{{produce}}}",
                "Rewrite: BoxJs",
                "Script Hub: 前端",
            ),
            required_mitm_hosts=("sub.store", "script.hub", "boxjs.com"),
            required_script_path_substrings=(
                "modules/source/scripts/github_com_sub-store-org_sub-store-1.min.js",
                "modules/source/scripts/raw_githubusercontent_com_f59ef7_script-hub.js",
            ),
        ),
        BundleSpec(
            "Weibo (PROMAX source)",
            Path(mb.WEIBO_LOCAL),
            mb.WEIBO_SOURCES,
        ),
    ]


def audit_promax_modules(root: Path) -> list[str]:
    issues: list[str] = []
    for rel in (
        "modules/surge/head_expanse/🚫 Universal Ad-Blocking Rules Dependency Component PROMAX (Kali-style).sgmodule",
    ):
        path = root / rel
        if not path.is_file():
            issues.append(f"PROMAX: missing {rel}")
            continue
        text = path.read_text(encoding="utf-8")
        if "%INSERT%" in text:
            issues.append(f"PROMAX: %INSERT% found in {path.name}")
    return issues


def main() -> int:
    root = Path(get_project_root())
    specs = _specs(root)
    Logger.section("Bundle completeness audit")
    all_issues: list[str] = []
    for spec in specs:
        Logger.info(f"Checking {spec.label}…")
        all_issues.extend(audit_bundle(spec))
    all_issues.extend(audit_promax_modules(root))

    if all_issues:
        Logger.error(f"Found {len(all_issues)} issue(s):")
        for item in all_issues:
            print(f"  • {item}")
        return 1

    Logger.success(f"All {len(specs)} bundle(s) + PROMAX passed completeness audit.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
